File: Graphs/graphs_dfs.go
/*
   This implementation represents a graph as a collection of nodes, where each node has a value and a list of
   neighboring nodes. The DFS algorithm is implemented recursively by starting from the given node, marking
   it as visited, and printing its value. Then, the algorithm recursively visits each of the neighboring
   nodes that have not been visited yet. The time complexity of DFS is O(V + E), where V is the number of
   vertices (nodes) in the graph and E is the number of edges between the vertices.

   Sample Input : graph = {
   'A': ['B', 'C'],
   'B': ['D', 'E'],
   'C': ['F'],
   'D': [],
   'E': ['F'],
   'F': []
   }
   Output : DFS Traversal: A B D E F C
   Here, we have a graph represented as an adjacency list, with nodes A, B, C, D, E, and F and their respective neighbors.
   The DFS traversal starts at node A and visits each node in the graph, outputting the final order in which the nodes were visited.

   The time complexity of the DFS algorithm is O(V + E), where V is the number of vertices (nodes) and E is the number of
   edges in the graph. This is because the algorithm visits every vertex and every edge once.

   The space complexity of DFS depends on the maximum depth of the recursion stack. In the worst case, where the tree or
   graph is completely unbalanced, the recursion stack can reach a depth of O(n), where n is the number of nodes in the graph.
   Therefore, the space complexity of DFS is O(n).

*/
package main

import "fmt"

// Node represents a single node in the graph
type Node struct {
    value int
    visited bool
    neighbors []*Node
}

// DFS traverses the graph starting from the given node using Depth First Search
func DFS(node *Node) {
    // Mark the current node as visited
    node.visited = true

    // Print the value of the current node
    fmt.Printf("%d ", node.value)

    // Visit each of the neighboring nodes
    for _, neighbor := range node.neighbors {
        // If the neighbor has not been visited yet, visit it recursively
        if !neighbor.visited {
            DFS(neighbor)
        }
    }
}

func main() {
    // Create the nodes of the graph
    node1 := &Node{value: 1}
    node2 := &Node{value: 2}
    node3 := &Node{value: 3}
    node4 := &Node{value: 4}
    node5 := &Node{value: 5}

    // Add the neighbors for each node
    node1.neighbors = []*Node{node2, node3}
    node2.neighbors = []*Node{node1, node4, node5}
    node3.neighbors = []*Node{node1, node5}
    node4.neighbors = []*Node{node2}
    node5.neighbors = []*Node{node2, node3}

    // Start the DFS traversal from node1
    DFS(node1)
}

File: Graphs/Graphs_dfs.js
// Takes in a graph object and a starting node
function dfs(graph, startNode) {
    // Initialize an empty visited object
    const visited = {}

    // Call the dfsHelper function with the starting node, the visited object, and the graph object
    dfsHelper(startNode, visited, graph)
}

// Recursive helper function for the DFS algorithm
// Takes in a current node, a visited object, and a graph object
function dfsHelper(node, visited, graph) {
    // Mark the current node as visited by adding it to the visited object
    visited[node] = true

    // Print the current node (or perform some other action)
    console.log(node)

    // Get the adjacent nodes of the current node from the graph object
    const adjacentNodes = graph[node]

    // Iterate over the adjacent nodes
    for (let i = 0; i < adjacentNodes.length; i++) {
        // Get the node of the current adjacent node
        const adjacentNode = adjacentNodes[i].node

        // If the adjacent node has not been visited yet, recursively call dfsHelper with the adjacent node
        if (!visited[adjacentNode]) {
            dfsHelper(adjacentNode, visited, graph)
        }
    }
}

// SAMPLE USE CASE
// The graph will be represented in form of an adjacency list.
/* 
    The graph will be in form of an object where each key represents the nodes and each value will be an array of the neighbors of the node.

    The array in values will be an object array, where each object has the node, which is the neighbor, and weight, which is the distance to that neighbor from the current node.
*/

const graph = {
    0: [{ node: 1, weight: 10 }],
    1: [
        { node: 0, weight: 10 },
        { node: 2, weight: 10 },
        { node: 3, weight: 5 },
    ],
    2: [{ node: 1, weight: 10 }],
    3: [
        { node: 1, weight: 5 },
        { node: 4, weight: 10 },
        { node: 5, weight: 10 },
    ],
    4: [
        { node: 3, weight: 10 },
        { node: 5, weight: 10 },
    ],
    5: [
        { node: 3, weight: 10 },
        { node: 4, weight: 10 },
    ],
}

dfs(graph, 0)